### PR TITLE
ci: add action to build and release binaries

### LINKED ISSUE #597: Use action to upload CI artifacts for `master` builds of svd2rust and attach them to a pre-release

### PR DESCRIPTION
**Pull request by duskmoon314**
- Execute when new commit in master and attach to pre-release
- Execute when new version tag and attach to release
- Allow manual execution

close #597

### CODE REVIEW

**Comment by duskmoon314** [comment]
Haven't written CHANGELOG yet.

An example of the result is [here](https://github.com/duskmoon314/svd2rust/releases/tag/nightly). Maybe there is no need to build daily. Just copy the idea from [rust-analyzer](https://github.com/rust-lang/rust-analyzer/releases/tag/nightly)

**Comment by burrbull** [comment]
> * Execute daily and attach to pre-release `nightly`
What sense of daily builds?

>     * Allow manual execution
How this works?

**Comment by duskmoon314** [comment]
> What sense of daily builds?

Execute every day at 00:00 UTC

> > ```
> > * Allow manual execution
> > ```
> 
> How this works?

I add `workflow_dispatch` to the trigger of action. So there is a button on the page of actions that will trigger once it is clicked.

image

Maybe not necessary either. I add this in order to ease the usage at first but it seems not very useful actually.

**Comment by burrbull** [comment]
> Execute every day at 00:00 UTC

This is not `linux` or `rustc` repo with hundred commits a day. What sense of daily builds?

**Comment by duskmoon314** [comment]
> > Execute every day at 00:00 UTC
> 
> This is not `linux` or `rustc` repo with hundred commits a day. What sense in daily builds?

You are right. I also think this is unnecessary.

**Comment by duskmoon314** [comment]
Remove daily build trigger.

**Comment by burrbull** [comment]
update changelog, please

**Comment by burrbull** [comment]
rebase on master, please

**Review by burrbull — APPROVED** [review]

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            suffix: ""
+          - os: macos-latest
+            target: x86_64-apple-darwin
+            suffix: ""
+          - os: macos-latest
+            target: aarch64-apple-darwin
+            suffix: ""
+          - os: windows-latest
+            target: x86_64-pc-windows-msvc
+            suffix: ".exe"
```
Comment (Emilgardis) [inline .github/workflows/release.yml:26]: the uploaded files for mac and linux should be compressed

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+          prerelease: true
+          name: "Pre-release"
+          tag_name: "pre-release"
+          files: |
+            artifacts/**/*
+      - uses: softprops/action-gh-release@v1
+        if: startsWith(github.ref, 'refs/tags/v')
+        with:
+          files: |
+            artifacts/**/*
```
Comment (Emilgardis) [inline .github/workflows/release.yml:72]: this should include current changelog

**Comment by burrbull** [comment]
Also title should include date and commit

File: .github/workflows/release.yml (hunk 1 above)
Comment (duskmoon314) [reply .github/workflows/release.yml:26]: You mean using `gzip -c svd2rust > svd2rust-<target>.gz` and upload that?

**Pull request merged**